src/Aspire.Dashboard/Components/Pages/TraceDetail.razor.cs

```diff
@@ -93,6 +94,43 @@ private ValueTask<GridItemsProviderResult<SpanWaterfallViewModel>> GetData(GridI
             Items = page.ToList(),
             TotalItemCount = visibleSpanWaterfallViewModels.Count
         });
+
+        // Return spans where any child span matches the filter.
+        List<SpanWaterfallViewModel> GetFilteredSpanWaterfalls(List<SpanWaterfallViewModel> viewModels)
```

Comment:
Could you split this method out from the page, maybe put it on `SpanWaterfallViewModel`, and then write some unit tests for each scenario you mentioned on the page.